### PR TITLE
Ghost, flying, and floor enemy updates. Floor enemy abstractification.

### PR DESCRIPTION
**Pull request by MattLiu-mygit**
Ghost enemy:
1. Removed patrolling feature and spawnpoints. Instead, implemented player tracking. The ghost now floats spoopyly towards the player and goes in and out of reality(disappears and reappears). Every time it disappears, it reappears at a random location around and above the player and floats towards at the player. Now it'll be more difficult and spoopy to deal with.

Flying:
1. Renamed to helicopter enemy. As aside from basic enemy characteristics flying and ghost enemy AIs differed greatly from one another, there seemed to be barely any abstractification possible. I made flying enemies distinct by renaming them helicopter enemies. 
    AI pattern Difference:
        - Ghost AI patterns is player chasing.
        - Helicopter AI patterns is patrolling.

2. However, as helicopter and ghost enemies are both flying enemies, I put them as separate categories of flying enemies.

3. Implemented a deceleration mechanic using the derivative of an s-curve to make the helicopter patrolling look a bit better. I recommend larger patrol cycles and faster speeds to make it look better. I also implemented a sprite rotation to make it look like the helicopter is trying to slow like in real life helicopters.(check it out, maybe with like SPEED = 100 and bounds at 100 or higher)

Floor enemies:
1. Abstractified floor enemies. Now it's pretty simple to make different kinds of floor enemies. I basically set it so that all floor enemies
    - Follow player mouse
    - Slide smoothly on the floor

2. Renamed the current unabstractified implementation of floor enemies into "sticky enemies" which inherits from floor enemies. Implemented it to stick to players and slow them to 25% of their movement speeds. Very annoying. 

Abstractification:
1. Refer to floor enemies.

2. Wanted to check in with you to see if abstractification of floor enemy is good enough.

3. Walking enemies is already pretty abstract. Stationary enemies only needs to inherit from normal enemy, there's no need to make a separate scene for them. Jumping enemies is already pretty abstract. Future enemies to make just need to add on top of jumping enemies. Current flying enemies are so different that maybe we can split AI patterns into patrolling(helicopter) and player tracking(ghost). Is this a good idea?
    Basically this could be the file structure:
        - flying
            - patrolling
                - helicopter
                - etc.
            - player tracking
                - ghost
                - etc.

### CODE REVIEW

**Comment by MattLiu-mygit** [comment]
Sorry, the indentation apparently didn't show up as well as I liked xD. Here is what I was actually thinking:
>    Basically this could be the file structure:
>    - flying
>    - - -  patrolling
>    - - - - - helicopter
>    - - - - - etc.
>    - - - player tracking
>    - - - - - ghost
>    - - - - - etc.

File: entities/enemies/floor/StickyEnemy/StickyEnemy.gd [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+var main_instances = ResourceLoader.main_instances
+var attached_to_player = false
+
+
+func _physics_process(delta) -> void:
+	var player = main_instances.player
+	
+	if attached_to_player:
+		motion.x = 0
+		player.motion.x *= SLOW_PERCENT
```
Comment (JJamesWWang) [inline entities/enemies/floor/StickyEnemy/StickyEnemy.gd:12]: What happens if the player jumps? Do they get unattached?

File: entities/enemies/flying/HelicopterEnemy/HelicopterEnemy.gd [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+	motion.x *= deceleration
+	decelerate_rotation(deceleration)
+
+
+# A check to see if it's in the right direction
+func return_to_patrol_area() -> void:
+	# Needs to check for this as this function will be called on constantly if 
+	# the enemy ever gets knocked out of its patrol area. If it's int he right
+	# direction, no need to perform a patrol flip. if not in right direction, 
+	# it hasn't performed a patrol flip and will perform one.
```
Comment (JJamesWWang) [inline entities/enemies/flying/HelicopterEnemy/HelicopterEnemy.gd:56]: I don't think we need these comments btw, the one above it is fine

File: entities/enemies/flying/HelicopterEnemy/HelicopterEnemy.gd [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+	
+	# Starts decelerating halfway to the patrol flip. Checks for moving right and
+	# moving left.
+	if (to_right_bound < 0.5 * RIGHT_BOUND and FLYING_DIRECTION == DIRECTION.RIGHT):
+		decelerate(to_right_bound, to_left_bound)
+	
+	if (to_left_bound < 0.5 * LEFT_BOUND and FLYING_DIRECTION == DIRECTION.LEFT):
+		decelerate(to_right_bound, to_left_bound)
+		
+	# Gave some room so enemy doesn't decelerated too much before the flip.
```
Comment (JJamesWWang) [inline entities/enemies/flying/HelicopterEnemy/HelicopterEnemy.gd:93]: Typo

File: entities/enemies/flying/HelicopterEnemy/HelicopterEnemy.gd [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+func decelerate_calc(to_right_bound, to_left_bound) -> float:
+	# Basically modeling acceleration after the derivative of an s curve.
+	# multiplied by acceleration mod to adapt the function with the current
+	# speed.
+	var acceleration_mod
+	if FLYING_DIRECTION == 1:
+		acceleration_mod = motion.x/to_right_bound
+	else:
+		acceleration_mod = motion.x/to_left_bound
+	var deceleration = (4 * pow(2.71828, acceleration_mod/2))/pow((1 + pow(2.71828, acceleration_mod/2)), 2.0)
```
Comment (JJamesWWang) [inline entities/enemies/flying/HelicopterEnemy/HelicopterEnemy.gd:111]: Might wanna make some constants, not really sure what 4 is doing there, and 2.71 could probably be replaced by e

File: entities/enemies/flying/HelicopterEnemy/HelicopterEnemy.gd [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+	if rotation_degrees > -25 and sign(motion.x) == 1:
+		rotation_degrees -= 4*(deceleration - (1 - deceleration))
+	elif rotation_degrees < 25 and sign(motion.x) == -1:
+		rotation_degrees += 4*(deceleration + (1 - deceleration))
+
+
+func _on_PatrolTimer_timeout() -> void:
+	motion.x = SPEED * FLYING_DIRECTION
+	rotation_degrees = FLYING_DIRECTION * 30
+	decelerating = false
```
Comment (JJamesWWang) [inline entities/enemies/flying/HelicopterEnemy/HelicopterEnemy.gd:127]: Can't wait to see this in action, looks really cool

File: entities/enemies/flying/ghost/GhostEnemy.gd [hunk 6]
```diff
@@ -6,79 +6,50 @@ extends Enemy
 enum DIRECTION {LEFT = -1, RIGHT = 1}
```
Comment (JJamesWWang) [inline entities/enemies/flying/ghost/GhostEnemy.gd:6]: I'm glad you've made a distinction between GhostEnemy and FlyingEnemy's movements. I suggest moving GhostEnemy out of FlyingEnemy and putting it in a folder called "homing" (cuz it homes in on the player). Then you can separate the two and have a base FlyingEnemy that follows patterns similar to the HelicopterEnemy.

File: entities/enemies/flying/ghost/GhostEnemy.gd [hunk 7]
```diff
@@ -6,79 +6,50 @@ extends Enemy
...
+
+
+# Chases player
+func chase_player(player, delta) -> void:
+	var direction = (player.global_position - global_position).normalized()
+	motion.y += spoopy_y_mod * spoopy_timer.time_left
+	motion += direction * ACCELERATION * delta
+	motion = motion.clamped(SPEED)
+	sprite.flip_h = global_position > player.global_position
+	motion = move_and_slide(motion)
```
Comment (JJamesWWang) [inline entities/enemies/flying/ghost/GhostEnemy.gd:52]: Could we make it so that the GhostEnemy stops chasing the Player while the Player is pointing near them with the mouse? Like a Boo in Mario. You'd probably have to calculate a conic area from the Player to the mouse using some weird math. Alternatively, I might add an Area2D to the Player to indicate their point of view. Let me know what you think.

Comment (JJamesWWang) [reply entities/enemies/flying/ghost/GhostEnemy.gd:52]: Also, make sure the spoopy timer keeps ticking while the Player is looking at them so the Ghost will eventually fade away and reappear if the Player doesn't kill them.

File: entities/enemies/stationary/laser/LaserEnemy.tscn [hunk 8]
```diff
@@ -23,6 +25,9 @@ collision_mask = 8
 [node name="Collider" parent="." index="2"]
 shape = SubResource( 1 )
 
+[node name="Collider" parent="Hurtbox" index="0"]
+shape = SubResource( 2 )
+
 [node name="EnemyStats" parent="." index="4"]
 max_health = 3
```
Comment (JJamesWWang) [inline entities/enemies/stationary/laser/LaserEnemy.tscn:33]: I don't know how much you've played around with the LaserEnemy, but you'll notice that the laser can go past the Player and not damage them if they move the mouse too quickly. I think I have a rotate_slowly() method in LaserEnemy which I haven't implemented. If you could make the LaserEnemy rotate towards the mouse at an exported (int) angular speed, that'd be great.

Comment (JJamesWWang) [reply entities/enemies/stationary/laser/LaserEnemy.tscn:33]: Don't do this yet though, save it for the next pull request.  /\  \\/

Comment (JJamesWWang) [reply entities/enemies/stationary/laser/LaserEnemy.tscn:33]: Also, can you make a separation between a defeatable laser enemy and a nondefeatable one? I think a non-defeatable one would make a level much harder.

File: entities/enemies/Enemy.gd [hunk 9]
```diff
@@ -18,3 +22,11 @@ func _on_Hurtbox_hit(damage: int, _spot: Vector2) -> void:
...
 func _on_EnemyStats_enemy_died() -> void:
 	die()
+
+
+# VisibilityNotifier2Ds set so that enemies tracking players don't track the
+# player before the players see the enemies. Good for setting up levels.
+# Disconnected rn because there's a signal calling error, but may need to be 
+# connected in the future.
+func _on_VisibilityNotifier2D_screen_entered() -> void:
+	set_physics_process(true)
```
[the remark this answers is not in the document]
Comment (MattLiu-mygit) [reply entities/enemies/Enemy.gd:32]: Ah :o I see, I think misinterpreted the game as a side-scroller kinda thing where we don't want to activate enemies too early xD, but yea I'll remove them 👍

File: entities/enemies/floor/StickyEnemy/StickyEnemy.gd (hunk 1 above)
Comment (MattLiu-mygit) [reply entities/enemies/floor/StickyEnemy/StickyEnemy.gd:12]: Works like a charm!

File: entities/enemies/floor/FloorEnemy.gd [hunk 10]
```diff
@@ -11,10 +13,12 @@ func _ready() -> void:
 
 func _physics_process(_delta: float) -> void:
 	var facing_direction = int(sign(get_local_mouse_position().x))
+	
+	# Prevents jumping off cliffs
```
[the remark this answers is not in the document]
Comment (MattLiu-mygit) [reply entities/enemies/floor/FloorEnemy.gd:17]: On it! :o

File: entities/enemies/flying/HelicopterEnemy/HelicopterEnemy.gd (hunk 3 above)
Comment (MattLiu-mygit) [reply entities/enemies/flying/HelicopterEnemy/HelicopterEnemy.gd:93]: huh :o

File: entities/enemies/flying/HelicopterEnemy/HelicopterEnemy.gd (hunk 4 above)
Comment (MattLiu-mygit) [reply entities/enemies/flying/HelicopterEnemy/HelicopterEnemy.gd:111]: I'll add some constants. Couldn't find gdscript e, but I'll make a constant as a global variable. The 4 is there from finding the derivative of an s-curve(I think bringing the square from the bottom up for derivative).

File: entities/enemies/flying/HelicopterEnemy/HelicopterEnemy.gd (hunk 5 above)
Comment (MattLiu-mygit) [reply entities/enemies/flying/HelicopterEnemy/HelicopterEnemy.gd:127]: Idk xD too much unnecessary effort went into that xD

File: entities/enemies/stationary/laser/LaserEnemy.tscn (hunk 8 above)
Comment (MattLiu-mygit) [reply entities/enemies/stationary/laser/LaserEnemy.tscn:33]: I'll try to implement it after the other things! :o

File: entities/enemies/Enemy.gd (hunk 9 above)
[the remark this answers is not in the document]
Comment (JJamesWWang) [reply entities/enemies/Enemy.gd:32]: I thought about doing a side-scroller, but I think it would encourage a much campier playstyle, like killing enemies one by one. I don't want one-on-one encounters like that in normal levels.
